main/src/ca/uwaterloo/flix/language/phase/EntryPoint.scala

```diff
@@ -133,78 +137,90 @@ object EntryPoint {
+  private def checkEntryPointResult(defn: TypedAst.Def, root: TypedAst.Root, traitEnv: Map[Symbol.TraitSym, Ast.TraitContext])(implicit sctx: SharedContext, flix: Flix): Unit = defn match {
     case TypedAst.Def(sym, TypedAst.Spec(_, _, _, _, _, declaredScheme, _, declaredEff, _, _, _), _) =>
       val resultTpe = declaredScheme.base.arrowResultType
       val unitSc = Scheme.generalize(Nil, Nil, Type.Unit, RigidityEnv.empty)
       val resultSc = Scheme.generalize(Nil, Nil, resultTpe, RigidityEnv.empty)
 
-      // Check for [[IllegalEntryPointEffect]]
-      if (declaredEff != Type.Pure && (declaredEff != Type.IO && declaredEff != Type.NonDet && declaredEff != Type.Sys)) {
-        return Validation.toSoftFailure((),EntryPointError.IllegalEntryPointEff(sym, declaredEff, declaredEff.loc))
+      // Check for IllegalEntryPointEffect
+      if (isBadEntryPointEffect(declaredEff)) {
+        val error = EntryPointError.IllegalEntryPointEff(sym, declaredEff, declaredEff.loc)
+        sctx.errors.add(error)
       }
 
-      if (Scheme.equal(unitSc, resultSc, traitEnv, ListMap.empty)) { // TODO ASSOC-TYPES better eqEnv
-        // Case 1: XYZ -> Unit.
-        Validation.success(())
-      } else {
-        // Delay ToString resolution if main has return type unit for testing with lib nix.
-        val toString = root.traits(new Symbol.TraitSym(Nil, "ToString", SourceLocation.Unknown)).sym
-        if (TraitEnvironment.holds(Ast.TraitConstraint(Ast.TraitConstraint.Head(toString, SourceLocation.Unknown), resultTpe, SourceLocation.Unknown), traitEnv, root.eqEnv)) {
-          // Case 2: XYZ -> a with ToString[a]
-          Validation.success(())
-        } else {
-          // Case 3: Bad result type. Error.
-          Validation.toSoftFailure((), EntryPointError.IllegalEntryPointResult(sym, resultTpe, sym.loc))
-        }
+      // Case 1: XYZ -> Unit.
+      val isUnitResult = Scheme.equal(unitSc, resultSc, traitEnv, ListMap.empty) // TODO ASSOC-TYPES better eqEnv
+
+      // Case 2: XYZ -> a with ToString[a]
+      val toStringTrait = root.traits(new Symbol.TraitSym(Nil, "ToString", SourceLocation.Unknown)).sym
+      val hasToStringConstraint = TraitEnvironment.holds(Ast.TraitConstraint(Ast.TraitConstraint.Head(toStringTrait, SourceLocation.Unknown), resultTpe, SourceLocation.Unknown), traitEnv, root.eqEnv)
+
+      // Case 3: Bad result type. Error.
+      val isBadResultType = !isUnitResult && !hasToStringConstraint
+
+      if (isBadResultType) {
+        val error = EntryPointError.IllegalEntryPointResult(sym, resultTpe, sym.loc)
+        sctx.errors.add(error)
       }
   }
 
+  /**
+    * Returns `true` iff `declaredEff` is not `Pure`, `IO`, `NonDet` or `Sys`
+    */
+  private def isBadEntryPointEffect(declaredEff: Type) = {
+    declaredEff != Type.Pure && (declaredEff != Type.IO && declaredEff != Type.NonDet && declaredEff != Type.Sys)
```

Comment:
Out of scope for this PR, but this looks like it should be looking for "base effects".